Public/Src/Pips/Dll/Operations/Process.cs

```diff
@@ -765,6 +793,119 @@ public bool TryComputePipUniqueOutputHash(PathTable pathTable, out long pipUniqu
             return true;
         }
 
+        /// <summary>
+        /// Compute the intersection of the pip's dependencies and global affected SourceChangeAffectedOutputs of the build
+        /// </summary>
+        public ReadOnlyArray<string> GetChangeAffectedInputNames(PathTable pathTable)
+        {
+            var changeAffectedInputs = SourceChangeAffectedOutputFiles.Intersect(Dependencies.Select(d => d.Path)).Select(i => i.GetName(pathTable).ToString(pathTable.StringTable)).ToHashSet();
+
+            foreach (var file in SourceChangeAffectedOutputFiles)
+            {
+                foreach (var directory in DirectoryDependencies)
+                {
+                    if (directory.Path.TryGetRelative(pathTable, file, out var relativePath))
+                    {
+                        changeAffectedInputs.Add(file.GetName(pathTable).ToString(pathTable.StringTable));
+                    }
+                }
+            }
+
+            foreach (var outDir in SourceChangeAffectedOutputDirectories)
+            {
+                foreach (var inDir in DirectoryDependencies)
+                {
+                    if (outDir.TryGetRelative(pathTable, inDir.Path, out var relativePath))
+                    {
+                        FileUtilities.EnumerateDirectoryEntries(
+                            inDir.Path.ToString(pathTable), 
+                            true, 
+                            (dir, fileName, attributes) =>
+                            {
+                                if (attributes == FileAttributes.Archive)
+                                {
+                                    changeAffectedInputs.Add(fileName);
+                                }
+                            });
+                    }
+                }
+            }
+
+            return ReadOnlyArray<string>.FromWithoutCopy(changeAffectedInputs.ToArray());
+        }
+
+        /// <inheritdoc />
+        public override bool IsOutputAffectedBySourceChange(
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedFiles,
+            ReadOnlyArray<AbsolutePath> dynamicallyObservedEnumerations,
+            PathTable pathTable,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputFiles = null,
+            IReadOnlyCollection<AbsolutePath> sourceChangeAffectedOutputDirectroies = null)
+        {
+            // sourceChangeAffectedOutputFiles and sourceChangeAffectedOutputDirectroies 
+            // are initilized with the source change list from GBR.
+            // If it is null, that means no file or directory change for this build.
+            // Thus no output is affected by the change.
+            if (!sourceChangeAffectedOutputFiles.Any() && !sourceChangeAffectedOutputDirectroies.Any())

```

Comment:
This `if` condition will blow up with null reference exception if they are `null`.